### PR TITLE
Fix a few flaky suites

### LINKED ISSUE #4301: Fix flaky KeepAlive Tests
A number of the keep alive tests in CompilerServerApiTests are disabled due to flaky behavior in the lab.  Need to fix the flaky behavior and re-enable these suites.

### PR DESCRIPTION
**Pull request by jaredpar**
This change fixes a few of the flaky compiler server suites.  Fixing this required a bit of work as the flakiness was related to our use of cross process / named pipes in the actual tests.  I refactored the code to leverage interfaces more in order to more directly test the scenarios involved.

closes #4301

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @agocke, @TyOverby,

File: src/Compilers/Server/PortableServer/CoreClrCompilerServerHost.cs [hunk 1]
```diff
@@ -13,78 +13,25 @@
...
 {
-    internal sealed class CoreClrCompilerServerHost : ICompilerServerHost
+    internal sealed class CoreClrCompilerServerHost : CompilerServerHost
     {
         private readonly Func<string, MetadataReferenceProperties, PortableExecutableReference> _assemblyReferenceProvider = (path, properties) => new CachingMetadataReference(path, properties);
         private readonly IAnalyzerAssemblyLoader _analyzerAssemblyLoader = CoreClrAnalyzerAssemblyLoader.CreateAndSetDefault();
-        private readonly TcpListener _listener;
 
-        public IAnalyzerAssemblyLoader AnalyzerAssemblyLoader => _analyzerAssemblyLoader;
+        public override IAnalyzerAssemblyLoader AnalyzerAssemblyLoader => _analyzerAssemblyLoader;
```
Comment (agocke) [inline src/Compilers/Server/PortableServer/CoreClrCompilerServerHost.cs]: Readonly autoprop?

Comment (jaredpar) [inline src/Compilers/Server/PortableServer/CoreClrCompilerServerHost.cs]: Sure

**Comment by TyOverby** [comment]
Looks like this is breaking the CI build

http://dotnet-ci.cloudapp.net/job/roslyn_master_win_rel_unit64/315/consoleFull#-1107647460f8aae81d-4bec-45a1-8770-9e05a4fee5ca

**Pull request merged**